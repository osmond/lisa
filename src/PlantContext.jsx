import { createContext, useContext, useEffect, useState } from "react";
import initialPlants from "./plants.json";
import { useWeather } from "./WeatherContext.jsx";
import { getNextWateringDate } from "./utils/watering.js";
import { getWaterPlan, getSmartWaterPlan } from "./utils/waterCalculator.js";
import { cubicInchesToMl, mlToOz } from "./utils/volume.js";
import autoTag from "./utils/autoTag.js";

const PlantContext = createContext();

export const addBase = (url) => {
  if (!url) return url;
  if (/^https?:/.test(url) || url.startsWith("data:")) return url;
  const base = (process.env.VITE_BASE_PATH || "/").replace(/\/$/, "");
  if (url.startsWith(base)) return url;
  return `${base}${url.startsWith("/") ? "" : "/"}${url}`;
};

const mapPhoto = (photo) => {
  if (!photo) return photo;
  if (typeof photo === "string") {
    return { src: addBase(photo), caption: "", tags: [] };
  }
  return { ...photo, src: addBase(photo.src), tags: photo.tags || [] };
};

export function PlantProvider({ children }) {
  const [plants, setPlants] = useState(() => {
    const mapPlant = (p) => {
      const wp = p.waterPlan || { interval: 0 };
      const ml =
        wp.volume_ml !== undefined
          ? wp.volume_ml
          : cubicInchesToMl(wp.volume || 0);
      const oz =
        wp.volume_oz !== undefined ? wp.volume_oz : mlToOz(ml);
      const smart = p.smartWaterPlan
        ? {
            ...p.smartWaterPlan,
            volume_ml:
              p.smartWaterPlan.volume_ml !== undefined
                ? p.smartWaterPlan.volume_ml
                : cubicInchesToMl(p.smartWaterPlan.volume || 0),
            volume_oz:
              p.smartWaterPlan.volume_oz !== undefined
                ? p.smartWaterPlan.volume_oz
                : mlToOz(
                    p.smartWaterPlan.volume_ml !== undefined
                      ? p.smartWaterPlan.volume_ml
                      : cubicInchesToMl(p.smartWaterPlan.volume || 0),
                  ),
          }
        : null;
      return {
        ...p,
        image: addBase(p.image),
        photos: (p.photos || p.gallery || []).map(mapPhoto),
        careLog: (p.careLog || []).map((ev) => ({ ...ev, tags: ev.tags || [] })),
        diameter: p.diameter || 0,
        petSafe: !!p.petSafe,
        waterPlan: { interval: wp.interval || 0, volume_ml: Math.round(ml), volume_oz: Math.round(oz) },
        carePlan: p.carePlan || null,
        smartWaterPlan: smart,
        lastEvaluated: p.lastEvaluated || null,
      };
    };

    if (typeof localStorage !== "undefined") {
      const stored = localStorage.getItem("plants");
      if (stored) {
        try {
          return JSON.parse(stored).map(mapPlant);
        } catch {
          // fall through to initial plants
        }
      }
    }
    return initialPlants.map(mapPlant);
  });

  useEffect(() => {
    let ignore = false;
    async function load() {
      if (typeof fetch === 'undefined') return;
      try {
        const res = await fetch("/api/plants");
        if (!res.ok) return;
        const data = await res.json();
        if (!ignore) setPlants(data.map(mapPlant));
      } catch {
        // ignore network errors and keep local data
      }
    }
    load();
    return () => {
      ignore = true;
    };
  }, []);

  const weatherCtx = useWeather();
  const weather = { rainTomorrow: weatherCtx?.forecast?.rainfall || 0 };

  useEffect(() => {
    if (typeof localStorage !== "undefined") {
      localStorage.setItem("plants", JSON.stringify(plants));
    }
  }, [plants]);

  const evaluatePlans = (forecast) => {
    if (!forecast) return;
    setPlants((prev) =>
      prev.map((p) => {
        if (!p.name || !p.diameter) return p;
        const logs = (p.careLog || []).filter((l) => l.type === "Watered");
        const raw = getSmartWaterPlan(
          { name: p.name, diameter: p.diameter, light: p.light },
          forecast,
          logs,
        );
        const ml = cubicInchesToMl(raw.volume);
        const plan = {
          interval: raw.interval,
          reason: raw.reason,
          volume_ml: Math.round(ml),
          volume_oz: Math.round(mlToOz(ml)),
        };
        return {
          ...p,
          smartWaterPlan: plan,
          lastEvaluated: new Date().toISOString(),
        };
      }),
    );
  };

  useEffect(() => {
    const handler = (e) => evaluatePlans(e.detail);
    window.addEventListener("weather-updated", handler);
    return () => window.removeEventListener("weather-updated", handler);
  }, []);

  const [timelineNotes, setTimelineNotes] = useState(() => {
    if (typeof localStorage !== "undefined") {
      const stored = localStorage.getItem("timelineNotes");
      if (stored) {
        try {
          return JSON.parse(stored).map((n) => ({ ...n, tags: n.tags || [] }));
        } catch {
          // ignore
        }
      }
    }
    return [];
  });

  useEffect(() => {
    if (typeof localStorage !== "undefined") {
      localStorage.setItem("timelineNotes", JSON.stringify(timelineNotes));
    }
  }, [timelineNotes]);

  const addTimelineNote = async (text) => {
    const date = new Date().toISOString().slice(0, 10);
    const tags = await autoTag(text);
    setTimelineNotes((prev) => [...prev, { date, text, tags }]);
  };

  const logEvent = async (id, type, note = "") => {
    const date = new Date().toISOString().slice(0, 10);
    const tags = await autoTag(note);
    setPlants((prev) =>
      prev.map((p) =>
        p.id === id
          ? {
              ...p,
              careLog: [...(p.careLog || []), { date, type, note, tags }],
            }
          : p,
      ),
    );
  };

  const markWatered = async (id, note) => {
    const tags = await autoTag(note);
    const today = new Date().toISOString().slice(0, 10);
    setPlants((prev) =>
      prev.map((p) => {
        if (p.id === id) {
          const { date: nextStr, reason } = getNextWateringDate(today, weather);
          const newLog = [
            ...(p.careLog || []),
            { date: today, type: "Watered", note, tags },
          ];
          const raw = getSmartWaterPlan(
            { name: p.name, diameter: p.diameter, light: p.light },
            weatherCtx?.forecast,
            newLog.filter((l) => l.type === "Watered"),
          );
          const volMl = cubicInchesToMl(raw.volume);
          const plan = {
            interval: raw.interval,
            reason: raw.reason,
            volume_ml: Math.round(volMl),
            volume_oz: Math.round(mlToOz(volMl)),
          };
          return {
            ...p,
            careLog: newLog,
            lastWatered: today,
            nextWater: nextStr,
            nextWaterReason: reason,
            smartWaterPlan: plan,
            lastEvaluated: new Date().toISOString(),
          };
        }
        return p;
      }),
    );
  };

  const markFertilized = (id, note) => {
    setPlants((prev) =>
      prev.map((p) => {
        if (p.id === id) {
          const today = new Date().toISOString().slice(0, 10);
          let nextDate = new Date(today);
          if (p.lastFertilized && p.nextFertilize) {
            const last = new Date(p.lastFertilized);
            const next = new Date(p.nextFertilize);
            const diff = Math.round((next - last) / 86400000) || 30;
            nextDate.setDate(nextDate.getDate() + diff);
          } else {
            nextDate.setMonth(nextDate.getMonth() + 1);
          }
          return {
            ...p,
            lastFertilized: today,
            nextFertilize: nextDate.toISOString().slice(0, 10),
          };
        }
        return p;
      }),
    );
    logEvent(id, "Fertilized", note);
  };

  const addPlant = (plant) => {
    setPlants((prev) => {
      const nextId = prev.reduce((m, p) => Math.max(m, p.id), 0) + 1;

      const newPlant = {
        id: nextId,
        photos: [],
        careLog: [],
        diameter: 0,
        waterPlan: { interval: 0, volume_ml: 0, volume_oz: 0 },
        carePlan: null,
        ...plant,
      };
      if (typeof fetch !== 'undefined')
        fetch('/api/plants', {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify(newPlant),
        }).catch(() => {})
      return [...prev, newPlant];
    });
  };

  const updatePlant = (id, updates) => {
    setPlants((prev) =>
      prev.map((p) => {
        if (p.id !== id) return p;
        const next = { ...p, ...updates };
        if (Object.prototype.hasOwnProperty.call(updates, "diameter")) {
          const base = getWaterPlan(next.name, updates.diameter, next.light);
          const ml = cubicInchesToMl(base.volume);
          next.waterPlan = {
            interval: base.interval,
            volume_ml: Math.round(ml),
            volume_oz: Math.round(mlToOz(ml)),
          };
        }
        return next;
      }),
    );
    if (typeof fetch !== 'undefined')
      fetch(`/api/plants/${id}`, {
        method: 'PUT',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(updates),
      }).catch(() => {})
  };

  const removePlant = (id) => {
    setPlants((prev) => prev.filter((p) => p.id !== id));
    if (typeof fetch !== 'undefined')
      fetch(`/api/plants/${id}`, { method: 'DELETE' }).catch(() => {})
  };

  const restorePlant = (plant, index) => {
    setPlants((prev) => {
      const arr = [...prev];
      const pos = index != null ? index : arr.length;
      arr.splice(pos, 0, plant);
      return arr;
    });
  };

  const addPhoto = async (id, photo) => {
    const newPhoto = mapPhoto(photo);
    const tags = await autoTag(newPhoto.caption || "");
    newPhoto.tags = tags;
    setPlants((prev) =>
      prev.map((p) =>
        p.id === id ? { ...p, photos: [...(p.photos || []), newPhoto] } : p,
      ),
    );
    const form = new FormData();
    if (photo.file) form.append('photos', photo.file);
    if (typeof fetch !== 'undefined' && form.has('photos'))
      fetch(`/api/plants/${id}/photos`, { method: 'POST', body: form }).catch(
        () => {},
      );
  };

  const removePhoto = (id, index) => {
    setPlants((prev) =>
      prev.map((p) => {
        if (p.id === id) {
          const updated = (p.photos || []).filter((_, i) => i !== index);
          return { ...p, photos: updated };
        }
        return p;
      }),
    );
    if (typeof fetch !== 'undefined')
      fetch(`/api/plants/${id}/photos/${index}`, { method: 'DELETE' }).catch(() => {});
  };

  return (
    <PlantContext.Provider
      value={{
        plants,
        markWatered,
        markFertilized,
        logEvent,
        addPlant,
        updatePlant,
        removePlant,
        restorePlant,
        addPhoto,
        removePhoto,
        timelineNotes,
        addTimelineNote,
      }}
    >
      {children}
    </PlantContext.Provider>
  );
}

export const usePlants = () => useContext(PlantContext);
