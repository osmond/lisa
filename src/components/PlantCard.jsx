import { Link, useNavigate } from 'react-router-dom'
import { useRef, useState } from 'react'
import { usePlants } from '../PlantContext.jsx'

export default function PlantCard({ plant }) {
  const navigate = useNavigate()
  const { markWatered, removePlant } = usePlants()
  const startX = useRef(0)
  const [deltaX, setDeltaX] = useState(0)

  const handleWatered = () => {
    const note = window.prompt('Optional note') || ''
    markWatered(plant.id, note)
  }

  const handlePointerDown = e => {
    startX.current = e.clientX ?? e.touches?.[0]?.clientX ?? 0
  }

  const handlePointerMove = e => {
    if (!startX.current) return
    const currentX = e.clientX ?? e.touches?.[0]?.clientX ?? 0
    setDeltaX(currentX - startX.current)
  }

  const handlePointerEnd = () => {
    const diff = deltaX
    setDeltaX(0)
    startX.current = 0
    if (diff > 75) {
      handleWatered()
    } else if (diff < -150) {
      removePlant(plant.id)
    } else if (diff < -75) {
      navigate(`/plant/${plant.id}/edit`)
    }
  }

  return (
    <div
      data-testid="card-wrapper"
      className="relative"
      onPointerDown={handlePointerDown}
      onPointerMove={handlePointerMove}
      onPointerUp={handlePointerEnd}
      onPointerCancel={handlePointerEnd}
      onTouchStart={handlePointerDown}
      onTouchMove={handlePointerMove}
      onTouchEnd={handlePointerEnd}
    >
      <div className="absolute inset-0 flex justify-between items-center px-4 pointer-events-none">
        <button
          onClick={handleWatered}
          className="bg-green-600 text-white px-2 py-1 rounded pointer-events-auto"
        >
          Water
        </button>
        <div className="flex gap-2">
          <button
            onClick={() => navigate(`/plant/${plant.id}/edit`)}
            className="bg-blue-600 text-white px-2 py-1 rounded pointer-events-auto"
          >
            Edit
          </button>
          <button
            onClick={() => removePlant(plant.id)}
            className="bg-red-600 text-white px-2 py-1 rounded pointer-events-auto"
          >
            Delete
          </button>
        </div>
      </div>
      <div
        className="p-4 border rounded-xl shadow-sm bg-white dark:bg-gray-800"
        style={{ transform: `translateX(${deltaX}px)`, transition: deltaX === 0 ? 'transform 0.2s' : 'none' }}
      >
        <Link to={`/plant/${plant.id}`} className="block mb-2">
          <img src={plant.image} alt={plant.name} loading="lazy" className="w-full h-48 object-cover rounded-md" />
          <h2 className="font-semibold text-lg mt-2">{plant.name}</h2>
        </Link>
        <p className="text-sm text-gray-600 dark:text-gray-400">Last watered: {plant.lastWatered}</p>
        <p className="text-sm text-green-700 font-medium">Next: {plant.nextWater}</p>
        <button
          onClick={handleWatered}
          className="mt-2 px-3 py-1 bg-green-100 text-green-700 rounded hover:bg-green-200 transition"
        >
          Watered
        </button>
      </div>
    </div>
  )
}
