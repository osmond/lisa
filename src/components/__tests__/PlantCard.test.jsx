import { render, screen, fireEvent } from '@testing-library/react'
import PlantCard from '../PlantCard.jsx'
import { MemoryRouter, useNavigate } from 'react-router-dom'
import { usePlants } from '../../PlantContext.jsx'

const navigateMock = jest.fn()
const markWatered = jest.fn()
const removePlant = jest.fn()

jest.mock('react-router-dom', () => {
  const actual = jest.requireActual('react-router-dom')
  return { ...actual, useNavigate: jest.fn() }
})

jest.mock('../../PlantContext.jsx', () => ({
  usePlants: jest.fn(),
}))

const usePlantsMock = usePlants

beforeEach(() => {
  navigateMock.mockClear()
  markWatered.mockClear()
  removePlant.mockClear()
  useNavigate.mockReturnValue(navigateMock)
  usePlantsMock.mockReturnValue({
    plants: [],
    markWatered,
    removePlant,
    updatePlant: jest.fn(),
    addPlant: jest.fn(),
    addPhoto: jest.fn(),
    removePhoto: jest.fn(),
  })
})

const plant = {
  id: 1,
  image: 'test.jpg',
  name: 'Aloe Vera',
  lastWatered: '2024-05-01',
  nextWater: '2024-05-07'
}

test('renders plant name', () => {
  render(
    <MemoryRouter>
      <PlantCard plant={plant} />
    </MemoryRouter>
  )
  expect(screen.getByText('Aloe Vera')).toBeInTheDocument()
})

test('water button triggers watering', () => {
  render(
    <MemoryRouter>
      <PlantCard plant={plant} />
    </MemoryRouter>
  )
  fireEvent.click(screen.getByText('Water'))
  expect(markWatered).toHaveBeenCalledWith(1)
})

test('edit button navigates to edit page', () => {
  render(
    <MemoryRouter>
      <PlantCard plant={plant} />
    </MemoryRouter>
  )
  fireEvent.click(screen.getByText('Edit'))
  expect(navigateMock).toHaveBeenCalledWith('/plant/1/edit')
})

test('delete button removes plant', () => {
  render(
    <MemoryRouter>
      <PlantCard plant={plant} />
    </MemoryRouter>
  )
  fireEvent.click(screen.getByText('Delete'))
  expect(removePlant).toHaveBeenCalledWith(1)
})
