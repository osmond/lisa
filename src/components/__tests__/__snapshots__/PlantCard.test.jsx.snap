// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot in dark mode 1`] = `
<div
  aria-label="Plant card for Aloe Vera"
  class="relative overflow-hidden group focus:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 focus-visible:ring-blue-500"
  data-testid="card-wrapper"
  tabindex="0"
>
  <div
    class="absolute inset-0 flex justify-between items-center px-4 transition
        pointer-events-none opacity-0
        group-hover:opacity-100 group-hover:pointer-events-auto group-hover:animate-fade-in-up
        group-focus-within:opacity-100 group-focus-within:pointer-events-auto group-focus-within:animate-fade-in-up
        "
  >
    <button
      class="bg-blue-600 hover:bg-blue-700 text-white px-2 py-1 rounded pointer-events-auto relative overflow-hidden flex items-center gap-1"
    >
      <svg
        aria-hidden="true"
        class="w-4 h-4"
        fill="currentColor"
        height="1em"
        viewBox="0 0 256 256"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="none"
          height="256"
          width="256"
        />
        <path
          d="M208,144c0-72-80-128-80-128S48,72,48,144a80,80,0,0,0,160,0Z"
          fill="none"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
        />
        <path
          d="M136.1,191.2a47.9,47.9,0,0,0,39.2-39.1"
          fill="none"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
        />
      </svg>
      Water
    </button>
    <div
      class="flex gap-2"
    >
      <button
        class="bg-blue-600 text-white px-2 py-1 rounded pointer-events-auto relative overflow-hidden font-body text-sm flex items-center gap-1"
      >
        <svg
          aria-hidden="true"
          class="w-4 h-4"
          fill="currentColor"
          height="1em"
          viewBox="0 0 256 256"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            fill="none"
            height="256"
            width="256"
          />
          <path
            d="M96,216H48a8,8,0,0,1-8-8V163.3a7.9,7.9,0,0,1,2.3-5.6l120-120a8,8,0,0,1,11.4,0l44.6,44.6a8,8,0,0,1,0,11.4Z"
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="16"
          />
          <line
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="16"
            x1="216"
            x2="96"
            y1="216"
            y2="216"
          />
          <line
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="16"
            x1="136"
            x2="192"
            y1="64"
            y2="120"
          />
        </svg>
        Edit
      </button>
      <button
        class="bg-red-600 text-white px-2 py-1 rounded pointer-events-auto relative overflow-hidden font-body text-sm flex items-center gap-1"
      >
        <svg
          aria-hidden="true"
          class="w-4 h-4"
          fill="currentColor"
          height="1em"
          viewBox="0 0 256 256"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            fill="none"
            height="256"
            width="256"
          />
          <line
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="16"
            x1="216"
            x2="40"
            y1="56"
            y2="56"
          />
          <line
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="16"
            x1="104"
            x2="104"
            y1="104"
            y2="168"
          />
          <line
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="16"
            x1="152"
            x2="152"
            y1="104"
            y2="168"
          />
          <path
            d="M200,56V208a8,8,0,0,1-8,8H64a8,8,0,0,1-8-8V56"
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="16"
          />
          <path
            d="M168,56V40a16,16,0,0,0-16-16H104A16,16,0,0,0,88,40V56"
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="16"
          />
        </svg>
        Delete
      </button>
    </div>
  </div>
  <div
    class="bg-white dark:bg-gray-700 rounded-2xl shadow-sm p-4 "
    style="transform: translateX(0px); transition: transform 0.2s;"
  >
    <a
      class="block mb-2"
      href="/plant/1"
    >
      <img
        alt="Aloe Vera"
        class="w-full h-48 object-cover rounded-xl"
        loading="lazy"
        src="test.jpg"
      />
      <h2
        class="font-bold text-[1.1rem] font-headline mt-2"
      >
        Aloe Vera
      </h2>
    </a>
    <p
      class="text-sm text-green-700 font-medium font-body"
    >
      Next: 
      2024-05-07
    </p>
    <button
      class="mt-2 px-3 py-1 bg-green-100 text-green-700 rounded hover:bg-green-200 transition relative overflow-hidden font-body text-sm flex items-center gap-1"
    >
      <svg
        aria-hidden="true"
        class="w-4 h-4"
        fill="currentColor"
        height="1em"
        viewBox="0 0 256 256"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="none"
          height="256"
          width="256"
        />
        <path
          d="M208,144c0-72-80-128-80-128S48,72,48,144a80,80,0,0,0,160,0Z"
          fill="none"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
        />
        <path
          d="M136.1,191.2a47.9,47.9,0,0,0,39.2-39.1"
          fill="none"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
        />
      </svg>
      Watered
    </button>
  </div>
</div>
`;
